lib/saluki-components/src/destinations/datadog/common/io.rs

```diff
@@ -0,0 +1,274 @@
+    {
+        let endpoints = config.endpoint().build_resolved_endpoints(maybe_refreshable_config)?;
+        let client = HttpClient::builder()
+            .with_request_timeout(config.request_timeout())
+            .with_retry_policy(config.retry().to_default_http_retry_policy())
+            .with_bytes_sent_counter(telemetry.bytes_sent().clone())
+            .with_endpoint_telemetry(metrics_builder, Some(endpoint_name))
+            .build()?;
+
+        Ok(Self {
+            config,
+            telemetry,
+            client,
+            endpoints,
+        })
+    }
+
+    pub async fn spawn(self) -> Handle<B> {
+        let (requests_tx, requests_rx) = mpsc::channel(32);
+        let (io_shutdown_tx, io_shutdown_rx) = oneshot::channel();
+
+        let Self {
+            config,
+            telemetry,
+            client,
+            endpoints,
+        } = self;
+
+        spawn_traced(run_io_loop(
+            requests_rx,
+            io_shutdown_tx,
+            config,
+            client,
+            telemetry,
+            endpoints,
+        ));
+
+        Handle {
+            requests_tx,
+            io_shutdown_rx,
+        }
+    }
+}
+
+async fn run_io_loop<S, B>(
+    mut requests_rx: mpsc::Receiver<(usize, Request<B>)>, io_shutdown_tx: oneshot::Sender<()>,
+    config: ForwarderConfiguration, service: S, telemetry: ComponentTelemetry,
+    resolved_endpoints: Vec<ResolvedEndpoint>,
+) where
+    S: Service<Request<B>, Response = hyper::Response<Incoming>> + Clone + Send + 'static,
+    S::Future: Send,
+    S::Error: Send + Into<BoxError>,
+    B: Body + Clone + Send + 'static,
+{
+    // Spawn an endpoint I/O task for each endpoint we're configured to send to.
+    //
+    // This gives us granularity at just the right level, since the places we expect there to be problems (and thus
+    // where concurrency can help make forward progress) are at the endpoint/API key level.
+    let mut endpoint_txs = Vec::new();
+    let task_barrier = Arc::new(Barrier::new(resolved_endpoints.len() + 1));
+
+    for resolved_endpoint in resolved_endpoints {
+        let endpoint_url = resolved_endpoint.endpoint().to_string();
+
+        let (endpoint_tx, endpoint_rx) = mpsc::channel(config.endpoint_buffer_size());
+        let task_barrier = Arc::clone(&task_barrier);
+        spawn_traced(run_endpoint_io_loop(
+            endpoint_rx,
+            task_barrier,
+            config.clone(),
+            service.clone(),
+            telemetry.clone(),
+            resolved_endpoint,
+        ));
+
+        endpoint_txs.push((endpoint_url, endpoint_tx));
+    }
+
+    // Listen for requests to forward, and send a copy of each one to each endpoint I/O task.
+    while let Some((metrics_count, request)) = requests_rx.recv().await {
+        for (endpoint_url, endpoint_tx) in &endpoint_txs {
+            if endpoint_tx.send((metrics_count, request.clone())).await.is_err() {
```

Comment:
Mostly just curious, I know this code was moved, but I assume we use some request body type with a cheap clone? Part of me wants to lose that as a generic arg if we aren't using it, but probably not worthwhile.

Reply:
We could/probably should lose it, since, yeah, we use a fixed type, `FrozenChunkedBytesBuffer`... which is indeed cheaply cloneable: the spiritual equivalent of `bytes::Bytes`.